In [1]:
import numpy as np
import pandas as pd

# Support Vector Machine

For Solving regressions problems

In [2]:
from sklearn.svm import SVC

### Use models from Linear Regression.ipynb file [Titanic test and validation models]

In [3]:
X_train = pd.read_pickle('X_train.pkl')
y_train = pd.read_pickle('y_train.pkl')

In [4]:
X_valid = pd.read_pickle('X_valid.pkl')
y_valid = pd.read_pickle('y_valid.pkl')

### Data Normalization

In [5]:
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()

In [6]:
X_train.head(10)

Sex        Age  SibSp  Parch      Fare  Embarked_C  Embarked_Q  \
PassengerId                                                                   
299            0  29.699118      0      0   30.5000           0           0   
885            0  25.000000      0      0    7.0500           0           0   
248            1  24.000000      0      2   14.5000           0           0   
479            0  22.000000      0      0    7.5208           0           0   
306            0   0.920000      1      2  151.5500           0           0   
317            1  24.000000      1      0   26.0000           0           0   
3              1  26.000000      0      0    7.9250           0           0   
406            0  34.000000      1      0   21.0000           0           0   
743            1  21.000000      2      2  262.3750           1           0   
261            0  29.699118      0      0    7.7500           0           1   

             Embarked_S  Pclass_1  Pclass_2  Pclass_3  
PassengerId                                            
299                   1         1         0         0  
885                   1         0         0         1  
248                   1         0         1         0  
479                   1         0         0         1  
306                   1         1         0         0  
317                   1         0         1         0  
3                     1         0         0         1  
406                   1         0         1         0  
743                   0         1         0         0  
261                   0         0         0         1

In [7]:
X_train.describe()

Sex         Age       SibSp       Parch        Fare  Embarked_C  \
count  668.000000  668.000000  668.000000  668.000000  668.000000   668.00000   
mean     0.342814   29.476233    0.553892    0.372754   32.179397     0.17515   
std      0.475006   13.009962    1.185279    0.795588   51.604012     0.38038   
min      0.000000    0.420000    0.000000    0.000000    0.000000     0.00000   
25%      0.000000   22.000000    0.000000    0.000000    7.925000     0.00000   
50%      0.000000   29.699118    0.000000    0.000000   14.400000     0.00000   
75%      1.000000   35.000000    1.000000    0.000000   30.500000     0.00000   
max      1.000000   80.000000    8.000000    6.000000  512.329200     1.00000   

       Embarked_Q  Embarked_S    Pclass_1    Pclass_2    Pclass_3  
count  668.000000  668.000000  668.000000  668.000000  668.000000  
mean     0.086826    0.738024    0.227545    0.211078    0.561377  
std      0.281791    0.440039    0.419561    0.408379    0.496590  
min      0.000000    0.000000    0.000000    0.000000    0.000000  
25%      0.000000    0.000000    0.000000    0.000000    0.000000  
50%      0.000000    1.000000    0.000000    0.000000    1.000000  
75%      0.000000    1.000000    0.000000    0.000000    1.000000  
max      1.000000    1.000000    1.000000    1.000000    1.000000

#### Using Describe() method biggest part of data
#### is 0,... and 1,...
#### Some columns contains 29,.... 668,... and so on
#### Need to convert them into 0,... and 1,... using SCALE
value > 0.5 = 1, else = 0

In [8]:
cols_for_scale = ['Age', 'SibSp', 'Parch', 'Fare']

In [9]:
X_train[cols_for_scale] = scaler.fit_transform(X_train[cols_for_scale])

### Check Results

In [10]:
X_train.head(10)

Sex       Age  SibSp     Parch      Fare  Embarked_C  Embarked_Q  \
PassengerId                                                                     
299            0  0.367921  0.000  0.000000  0.059532           0           0   
885            0  0.308872  0.000  0.000000  0.013761           0           0   
248            1  0.296306  0.000  0.333333  0.028302           0           0   
479            0  0.271174  0.000  0.000000  0.014680           0           0   
306            0  0.006283  0.125  0.333333  0.295806           0           0   
317            1  0.296306  0.125  0.000000  0.050749           0           0   
3              1  0.321438  0.000  0.000000  0.015469           0           0   
406            0  0.421965  0.125  0.000000  0.040989           0           0   
743            1  0.258608  0.250  0.333333  0.512122           1           0   
261            0  0.367921  0.000  0.000000  0.015127           0           1   

             Embarked_S  Pclass_1  Pclass_2  Pclass_3  
PassengerId                                            
299                   1         1         0         0  
885                   1         0         0         1  
248                   1         0         1         0  
479                   1         0         0         1  
306                   1         1         0         0  
317                   1         0         1         0  
3                     1         0         0         1  
406                   1         0         1         0  
743                   0         1         0         0  
261                   0         0         0         1

In [11]:
X_train.describe()

Sex         Age       SibSp       Parch        Fare  Embarked_C  \
count  668.000000  668.000000  668.000000  668.000000  668.000000   668.00000   
mean     0.342814    0.365120    0.069237    0.062126    0.062810     0.17515   
std      0.475006    0.163483    0.148160    0.132598    0.100724     0.38038   
min      0.000000    0.000000    0.000000    0.000000    0.000000     0.00000   
25%      0.000000    0.271174    0.000000    0.000000    0.015469     0.00000   
50%      0.000000    0.367921    0.000000    0.000000    0.028107     0.00000   
75%      1.000000    0.434531    0.125000    0.000000    0.059532     0.00000   
max      1.000000    1.000000    1.000000    1.000000    1.000000     1.00000   

       Embarked_Q  Embarked_S    Pclass_1    Pclass_2    Pclass_3  
count  668.000000  668.000000  668.000000  668.000000  668.000000  
mean     0.086826    0.738024    0.227545    0.211078    0.561377  
std      0.281791    0.440039    0.419561    0.408379    0.496590  
min      0.000000    0.000000    0.000000    0.000000    0.000000  
25%      0.000000    0.000000    0.000000    0.000000    0.000000  
50%      0.000000    1.000000    0.000000    0.000000    1.000000  
75%      0.000000    1.000000    0.000000    0.000000    1.000000  
max      1.000000    1.000000    1.000000    1.000000    1.000000

### Repeat the same with X_valid

In [12]:
X_valid[cols_for_scale] = scaler.fit_transform(X_valid[cols_for_scale])

In [13]:
X_valid.describe()

Sex         Age       SibSp       Parch        Fare  Embarked_C  \
count  223.000000  223.000000  223.000000  223.000000  223.000000   223.00000   
mean     0.381166    0.420932    0.107623    0.081614    0.123024     0.22870   
std      0.486766    0.185043    0.200417    0.167582    0.166092     0.42094   
min      0.000000    0.000000    0.000000    0.000000    0.000000     0.00000   
25%      0.000000    0.315947    0.000000    0.000000    0.030094     0.00000   
50%      0.000000    0.411417    0.000000    0.000000    0.058107     0.00000   
75%      1.000000    0.501211    0.250000    0.200000    0.119414     0.00000   
max      1.000000    1.000000    1.000000    1.000000    1.000000     1.00000   

       Embarked_Q  Embarked_S    Pclass_1    Pclass_2    Pclass_3  
count  223.000000  223.000000  223.000000  223.000000  223.000000  
mean     0.085202    0.686099    0.286996    0.192825    0.520179  
std      0.279810    0.465121    0.453377    0.395404    0.500717  
min      0.000000    0.000000    0.000000    0.000000    0.000000  
25%      0.000000    0.000000    0.000000    0.000000    0.000000  
50%      0.000000    1.000000    0.000000    0.000000    1.000000  
75%      0.000000    1.000000    1.000000    0.000000    1.000000  
max      1.000000    1.000000    1.000000    1.000000    1.000000

### This result can be stored in to file

In [14]:
import joblib as jb

In [15]:
jb.dump(scaler, 'min_max_scaler.pkl')

['min_max_scaler.pkl']

#### Read file

In [16]:
scaler = jb.load('min_max_scaler.pkl')





# SVC





In [17]:
clf = SVC()

In [18]:
clf.fit(X_train,y_train)

SVC()

In [19]:
y_pred = clf.predict(X_valid)

In [20]:
y_pred_train = clf.predict(X_train)

### Validate Accurancy

In [21]:
from sklearn.metrics import accuracy_score

In [22]:
accuracy_score(y_valid, y_pred)

0.8026905829596412

In [23]:
accuracy_score(y_train, y_pred_train)

0.8158682634730539

# Tuning SVC model

c_vals = 1, pick random numbers 29 times from -2 to 5

In [24]:
c_vals = np.logspace(-2, 5, 29)
c_vals

array([1.00000000e-02, 1.77827941e-02, 3.16227766e-02, 5.62341325e-02,
       1.00000000e-01, 1.77827941e-01, 3.16227766e-01, 5.62341325e-01,
       1.00000000e+00, 1.77827941e+00, 3.16227766e+00, 5.62341325e+00,
       1.00000000e+01, 1.77827941e+01, 3.16227766e+01, 5.62341325e+01,
       1.00000000e+02, 1.77827941e+02, 3.16227766e+02, 5.62341325e+02,
       1.00000000e+03, 1.77827941e+03, 3.16227766e+03, 5.62341325e+03,
       1.00000000e+04, 1.77827941e+04, 3.16227766e+04, 5.62341325e+04,
       1.00000000e+05])

Now collect some data about training and validating

In [ ]:
accuracy_valid = []
accuracy_train = []
for val in c_vals:
    clf = SVC(C = val)
    clf.fit(X_train, y_train)
    y_pred = clf.predict(X_valid)
    y_pred_train = clf.predict(X_train)
    acc_valid = accuracy_score(y_valid, y_pred)
    acc_train = accuracy_score(y_train, y_pred_train)
    accuracy_valid.append(acc_valid)
    accuracy_train.append(acc_train)
    print('C = {} \n\tacc_valid = {} \n\tacc_train = {}\n'.format(val, acc_valid, acc_train))

C = 0.01 
	acc_valid = 0.600896860986547 
	acc_train = 0.6212574850299402

C = 0.01778279410038923 
	acc_valid = 0.600896860986547 
	acc_train = 0.6212574850299402

C = 0.03162277660168379 
	acc_valid = 0.7802690582959642 
	acc_train = 0.7889221556886228

C = 0.05623413251903491 
	acc_valid = 0.7847533632286996 
	acc_train = 0.7979041916167665

C = 0.1 
	acc_valid = 0.8116591928251121 
	acc_train = 0.812874251497006

C = 0.1778279410038923 
	acc_valid = 0.8071748878923767 
	acc_train = 0.812874251497006

C = 0.31622776601683794 
	acc_valid = 0.8071748878923767 
	acc_train = 0.812874251497006

C = 0.5623413251903491 
	acc_valid = 0.8071748878923767 
	acc_train = 0.812874251497006

C = 1.0 
	acc_valid = 0.8026905829596412 
	acc_train = 0.8158682634730539

C = 1.7782794100389228 
	acc_valid = 0.7982062780269058 
	acc_train = 0.8158682634730539

C = 3.1622776601683795 
	acc_valid = 0.8071748878923767 
	acc_train = 0.8158682634730539

C = 5.623413251903491 
	acc_valid = 0.8071748878923767 


# Show result in graph

In [ ]:
import matplotlib.pyplot as plt
%config InlineBackend.figure_format = 'svg'
%matplotlib inline

In [ ]:
plt.plot(c_vals, accuracy_valid)
plt.plot(c_vals, accuracy_train)
plt.xlabel('C value graph')
plt.ylabel('Accuracy')
plt.legend(['valid', 'train'])